File: java/src/main/java/com/aaronicsubstances/shrewd/evolver/logging/CustomLoggerFacade.java
// tag: 20210724T0000
package com.aaronicsubstances.shrewd.evolver.logging;

import java.util.function.Function;
import java.util.function.Supplier;

public class CustomLoggerFacade {
    private static final NoOpLogger DEFAULT_LOGGER = new NoOpLogger();
    private static Function<Object[], CustomLogger> loggerFactoryFunction;

    public static Function<Object[], CustomLogger> getLoggerFactoryFunction() {
        return loggerFactoryFunction;
    }

    public static void setLoggerFactoryFunction(Function<Object[], CustomLogger> value) {
        loggerFactoryFunction = value;
    }

    public static CustomLogger getLogger(Object classOrName) {
        if (loggerFactoryFunction != null) {
            return loggerFactoryFunction.apply(new Object[] { classOrName });
        }
        else {
            return DEFAULT_LOGGER;
        }
    }

    public static class NoOpLogger implements CustomLogger
    {
        public boolean isLogLevelEnbled(CustomLogLevel level) {
            return false;
        }

        public void log(CustomLogLevel level, String message, Object... args) {
        }

        public void log(CustomLogLevel level, Throwable error, String message, Object... args) {
        }

        public void log(CustomLogLevel level, CustomLogEvent logEvent) {
        }

        public void log(CustomLogLevel level, Supplier<CustomLogEvent> logEventSupplier) {
        }
    }
}

File: csharp/ShrewdEvolver/Logging/CustomLoggerFacade.cs
// tag: 20210724T0000
using System;
using System.Collections.Generic;
using System.Text;

namespace AaronicSubstances.ShrewdEvolver.Logging
{
    public static class CustomLoggerFacade
    {
        private static readonly NoOpLogger DefaultLogger = new NoOpLogger();
        public static Func<object[], ICustomLogger> LoggerFactoryFunction { get; set; }

        public static ICustomLogger GetLogger(object typeOrName)
        {
            if (LoggerFactoryFunction != null)
            {
                return LoggerFactoryFunction.Invoke(new object[] { typeOrName });
            }
            else
            {
                return DefaultLogger;
            }
        }

        public class NoOpLogger : ICustomLogger
        {
            public bool IsLogLevelEnbled(CustomLogLevel level)
            {
                return false;
            }

            public void Log(CustomLogLevel level, string message, params object[] args)
            {
            }

            public void Log(CustomLogLevel level, Exception error, string message, params object[] args)
            {
            }

            public void Log(CustomLogLevel level, CustomLogEvent logEvent)
            {
            }

            public void Log(CustomLogLevel level, Func<CustomLogEvent> logEventSupplier)
            {
            }
        }
    }
}
